src/goto-instrument/contracts/dynamic-frames/dfcc_instrument_loop.cpp

```diff
@@ -0,0 +1,717 @@
+    initial_invariant_assignment.add_source_location() = loop_head_location;
+    converter.goto_convert(
+      initial_invariant_assignment, pre_loop_head_instrs, language_mode);
+  }
+
+  {
+    // Create a temporary variable to track base case vs inductive case
+    // instrumentation of the loop.
+    // in_base_case = true;
+    pre_loop_head_instrs.add(
+      goto_programt::make_decl(in_base_case, loop_head_location));
+    pre_loop_head_instrs.add(
+      goto_programt::make_assignment(in_base_case, true_exprt{}));
+  }
+
+  {
+    // create and populate the write set
+    // DECL loop_write_set
+    // DECL addr_of_loop_write_set
+    // ASSIGN write_set_ptr := address_of(write_set)
+    // CALL __CPROVER_contracts_write_set_create(write_set_ptr,
+    //        contracts_assigns_size, contracts_assigns_frees_size,
+    //        assume_require_ctx, assert_require_ctx, assume_ensures_ctx,
+    //        assert_ensures_ctx, allow_allocate, allow_deallocate);
+    pre_loop_head_instrs.add(
+      goto_programt::make_decl(loop_write_set, loop_head_location));
+    pre_loop_head_instrs.add(
+      goto_programt::make_decl(addr_of_loop_write_set, loop_head_location));
+    pre_loop_head_instrs.add(goto_programt::make_assignment(
+      addr_of_loop_write_set,
+      address_of_exprt(loop_write_set),
+      loop_head_location));
+
+    code_function_callt call = library.write_set_create_call(
+      addr_of_loop_write_set,
+      from_integer(assigns.size(), size_type()),
+      from_integer(0, size_type()),
+      false_exprt(),
+      false_exprt(),
+      false_exprt(),
+      false_exprt(),
+      false_exprt(),
+      false_exprt(),
+      loop_head_location);
+    pre_loop_head_instrs.add(
+      goto_programt::make_function_call(call, loop_head_location));
+
+    pre_loop_head_instrs.destructive_append(assigns_instrs);
+  }
+
+  // goto HEAD;
+  pre_loop_head_instrs.add(
+    goto_programt::make_goto(loop_head, true_exprt{}, loop_head_location));
+
+  goto_function.body.destructive_insert(loop_head, pre_loop_head_instrs);
+  return history_var_map;
+}
+
+goto_programt::instructiont::targett
+dfcc_instrument_loopt::add_step_instructions(
+  const std::size_t loop_id,
+  const std::size_t cbmc_loop_id,
+  const irep_idt &function_id,
+  goto_functionst::goto_functiont &goto_function,
+  symbol_tablet &symbol_table,
+  goto_programt::targett loop_head,
+  goto_programt::targett loop_latch,
+  goto_programt &havoc_instrs,
+  const exprt &invariant,
+  const exprt::operandst &decreases_clauses,
+  const symbol_exprt &addr_of_loop_write_set,
+  const exprt &outer_write_set,
+  const symbol_exprt &initial_invariant,
+  const symbol_exprt &in_base_case,
+  const std::vector<symbol_exprt> &old_decreases_vars,
+  const irep_idt &language_mode)
+{
+  auto loop_head_location(loop_head->source_location());
+  dfcc_remove_loop_tags(loop_head_location);
+  goto_convertt converter(symbol_table, message_handler);
```

Comment:
I think the `assume(invariant)` block is the only one that actually needs this, please move it there.

Reply:
Correction: please move to just before the `assume(invariant)` block as the `decreases_clauses` case also needs it.